x/ecocredit/core/origin_tx.go

```diff
@@ -0,0 +1,30 @@
+package core
+
+import (
+	"regexp"
+
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+var reOriginTxId = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9 _\-]{0,127}$`)
+var reOriginTxSource = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9 _\-]{0,31}$`)
+
+func (o *OriginTx) Validate() error {
+	if !reOriginTxId.MatchString(o.Id) {
+		return sdkerrors.ErrInvalidRequest.Wrap("origin_tx.id must be at most 128 characters long, valid characters: alpha-numberic, space, '-' or '_'")
+	}
+
+	if !reOriginTxSource.MatchString(o.Source) {
+		return sdkerrors.ErrInvalidRequest.Wrap("origin_tx.source must be at most 32 characters long, valid characters: alpha-numberic, space, '-' or '_'")
+	}
+
+	if len(o.Contract) > 0 && !isValidEthereumAddress(o.Contract) {
```

Comment:
the `isValidEthereum` regexp catches the `>0` condition, so could simplify here

Reply:
So this is checking valid only if not empty. In some cases, we don't want contract to be required. For example, using mint batch credits when dynamically minting to a credit batch, which could be used as a manual bridge operation when bridging credits from a non-chain registry such as verra. https://github.com/regen-network/regen-ledger/pull/1206#discussion_r904122339

Reply:
We could make this required when a param is set to true (e.g. `contractRequired` or something) or if we make this always required we would only be supporting EVM-based chains wherever origin tx is being used.

Reply:
ahhh right - i keep forgetting about the manual case 😖